vta/hardware/chisel/src/main/scala/core/TensorAlu.scala

```diff
@@ -0,0 +1,316 @@
+  val fixme_alu_op = Cat(neg_shift, Mux(neg_shift, 0.U, dec.alu_op))
+  alu.io.opcode := fixme_alu_op
+  alu.io.acc_a.data.valid := io.acc.rd.data.valid & state === sReadTensorB
+  alu.io.acc_a.data.bits <> io.acc.rd.data.bits
+  alu.io.acc_b.data.valid := Mux(dec.alu_use_imm, tensorImm.data.valid, io.acc.rd.data.valid & state === sExe)
+  alu.io.acc_b.data.bits <> Mux(dec.alu_use_imm, tensorImm.data.bits, io.acc.rd.data.bits)
+
+  // acc_o
+  io.acc.wr.valid := alu.io.acc_y.data.valid
+  io.acc.wr.bits.idx := uop_dst
+  io.acc.wr.bits.data <> alu.io.acc_y.data.bits
+
+  // out
+  io.out.wr.valid := alu.io.out.data.valid
+  io.out.wr.bits.idx := uop_dst
+  io.out.wr.bits.data <> alu.io.out.data.bits
+  io.out.tieoffRead() // write-only
+
+  io.done := done
+
+  if (debug) {
+
+    when (state === sReadUop) {
+      printf("[TensorAlu] [uop] idx:%x\n", uop_idx)
+    }
+
+    when (state === sReadTensorA) {
+      printf("[TensorAlu] [uop] dst:%x src:%x\n", uop_dst, uop_src)
+    }
+
+    when (state === sIdle && io.start) {
+      when (dec.alu_op === ALU_OP_MIN) {
+        printf("[TensorAlu] [MIN] uop_begin:%x uop_end:%x lp[0]:%x lp[1]:%x dst[0]:%x dst[1]:%x src[0]:%x src[1]:%x useImm:%x imm:%x\n",
+          dec.uop_begin, dec.uop_end,
+	  dec.lp_0, dec.lp_1, dec.dst_0, dec.dst_1,
+          dec.src_0, dec.src_1, dec.alu_use_imm, dec.alu_imm)
+      } .elsewhen (dec.alu_op === ALU_OP_MAX) {
+        printf("[TensorAlu] [MAX] uop_begin:%x uop_end:%x lp[0]:%x lp[1]:%x dst[0]:%x dst[1]:%x src[0]:%x src[1]:%x useImm:%x imm:%x\n",
+          dec.uop_begin, dec.uop_end,
+          dec.lp_0, dec.lp_1, dec.dst_0, dec.dst_1,
+          dec.src_0, dec.src_1, dec.alu_use_imm, dec.alu_imm)
+      } .elsewhen (dec.alu_op === ALU_OP_ADD) {
+        printf("[TensorAlu] [ADD] uop_begin:%x uop_end:%x lp[0]:%x lp[1]:%x dst[0]:%x dst[1]:%x src[0]:%x src[1]:%x useImm:%x imm:%x\n",
+          dec.uop_begin, dec.uop_end,
+          dec.lp_0, dec.lp_1, dec.dst_0, dec.dst_1,
+          dec.src_0, dec.src_1, dec.alu_use_imm, dec.alu_imm)
+      } .elsewhen (dec.alu_op === ALU_OP_SHR) {
+        printf("[TensorAlu] [SHR] uop_begin:%x uop_end:%x lp[0]:%x lp[1]:%x dst[0]:%x dst[1]:%x src[0]:%x src[1]:%x useImm:%x imm:%x\n",
+          dec.uop_begin, dec.uop_end,
+          dec.lp_0, dec.lp_1, dec.dst_0, dec.dst_1,
+          dec.src_0, dec.src_1, dec.alu_use_imm, dec.alu_imm)
+      } .elsewhen (dec.alu_op === ALU_OP_SHL) {
+        printf("[TensorAlu] [SHL] uop_begin:%x uop_end:%x lp[0]:%x lp[1]:%x dst[0]:%x dst[1]:%x src[0]:%x src[1]:%x useImm:%x imm:%x\n",
+          dec.uop_begin, dec.uop_end,
+          dec.lp_0, dec.lp_1, dec.dst_0, dec.dst_1,
+          dec.src_0, dec.src_1, dec.alu_use_imm, dec.alu_imm)
+      }
+    }
+
+    alu.io.acc_a.data.bits.foreach { tensor =>
+      tensor.zipWithIndex.foreach { case(elem, i) =>
+        when (alu.io.acc_a.data.valid) {
+          printf("[TensorAlu] [a] i:%x val:%x\n", i.U, elem)
```

Comment:
Could we use meta-programming to generate all the debug output for each module?

Reply:
I think we can use string interpolation to "pretty print" the needed bundle. I just added that fix [here](https://github.com/dmlc/tvm/pull/3258/commits/16d77e05b69aef2c54ff323d5e75b41705d78b3d)

More info [here](https://github.com/freechipsproject/chisel3/wiki/Printing-in-Chisel)